api/envoy/config/filter/udp/dns_filter/v2alpha/dns_filter.proto

```diff
@@ -0,0 +1,86 @@
+syntax = "proto3";
+
+package envoy.config.filter.udp.dns_filter.v2alpha;
+
+import "envoy/api/v2/core/base.proto";
+
+import "google/protobuf/duration.proto";
+
+import "udpa/annotations/status.proto";
+
+import "udpa/annotations/migrate.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.config.filter.udp.dns_filter.v2alpha";
+option java_outer_classname = "DnsFilterProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_migrate).move_to_package =
+    "envoy.extensions.filter.udp.dns_filter.v3";
+option (udpa.annotations.file_status).work_in_progress = true;
+
+// [#protodoc-title: DNS Filter]
+// DNS Filter :ref:`configuration overview <config_udp_listener_filters_dns_filter>`.
+// [#extension: envoy.filters.udp_listener.dns_filter]
+
+// Configuration for the DNS filter.
+message DnsFilterConfig {
+  // This message contains a list of IP addresses returned for a query for a known name
+  message AddressList {
+    repeated string address = 1;
+  }
+
+  // This message type is extensible and can contain a list of addresses
+  // or dictate some other method for resolving the addresses for an
+  // endpoint
+  message DnsEndpoint {
+    oneof endpoint_config {
+      AddressList addresslist = 1;
+    }
+  }
+
+  message DnsVirtualDomain {
+    // The domain name for which Envoy will respond to query requests
+    string name = 1 [(validate.rules).string = {min_len: 2 well_known_regex: HTTP_HEADER_NAME}];
+
+    // The configuration containing the method to determine the address
+    // of this endpoint
+    DnsEndpoint endpoint = 2;
+
+    // Sets the TTL in dns answers from Envoy returned to the client
+    google.protobuf.Duration answer_ttl = 3 [(validate.rules).duration = {gt {}}];
+  }
+
+  // This message contains the configuration for the DNS Filter if populated
+  // from the control plane
+  message ControlPlaneConfig {
+    // Control how many times envoy makes an attempt to forward a query to
+    // an external server
+    uint32 external_retry_count = 1;
+
+    // Fully qualified domain names for which Envoy will respond to queries
+    repeated DnsVirtualDomain virtual_domains = 2 [(validate.rules).repeated = {min_items: 1}];
+
+    // This field serves to help Envoy determine whether a name belongs
+    // to a known resource or whether the query should be forwarded externally
+    repeated string known_domains = 3;
```

Comment:
Thanks for adding! 
A minor suggestion is to move this to field 2 and move the virtual_domains to field 3. Reasoning is that someone writing or reading the config can easily go over the list of known domains before going over a large list of actual mappings. Does that sound reasonable?

Reply:
I think it works better as it is since the known_domains is optional.   I prefer to keep the required fields before the optional ones.

Reply:
if the known_domains is empty, it means the dns filter is not authoritative for any domain. That means we are using the filter as a forwarder to send the queries to an external client.